optuna/_gp/_search_space.py

```diff
@@ -0,0 +1,162 @@
+from enum import IntEnum
+import math
+from typing import Any
+
+import numpy as np
+import scipy.stats.qmc
+
+import optuna
+
+
+class ParamType(IntEnum):
+    LINEAR = 0
+    LOG = 1
+    CATEGORICAL = 2
+
+
+@dataclass(frozen=True)
+class SearchSpace:
+    param_type: np.ndarray
+    bounds: np.ndarray
+    step: np.ndarray
+
+
+def untransform_one_param(
+    param_value: np.ndarray, param_type: ParamType, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    if param_type == ParamType.CATEGORICAL:
+        return param_value
+    else:
+        bounds2 = (bounds[0] - 0.5 * step, bounds[1] + 0.5 * step)
+        if param_type == ParamType.LOG:
+            bounds2 = (math.log(bounds2[0]), math.log(bounds2[1]))
+        param_value = param_value * (bounds2[1] - bounds2[0]) + bounds2[0]
+        if param_type == ParamType.LOG:
+            param_value = np.exp(param_value)
+        return param_value
+
+
+def transform_one_param(
+    param_value: np.ndarray, param_type: ParamType, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    if param_type == ParamType.CATEGORICAL:
+        return param_value
+    else:
+        bounds2 = (bounds[0] - 0.5 * step, bounds[1] + 0.5 * step)
+        if param_type == ParamType.LOG:
+            bounds2 = (math.log(bounds2[0]), math.log(bounds2[1]))
+            param_value = np.log(param_value)
+        param_value = (param_value - bounds2[0]) / (bounds2[1] - bounds2[0])
+        return param_value
+
+
+def round_one_transformed_param(
+    param_value: np.ndarray, param_type: ParamType, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    assert param_type != ParamType.CATEGORICAL
+    if step == 0.0:
+        return param_value
+
+    param_value = untransform_one_param(param_value, param_type, bounds, step)
+    param_value = (param_value - bounds[0] + 0.5 * step) // step * step + bounds[0]
+    param_value = transform_one_param(param_value, param_type, bounds, step)
+    return param_value
+
+
+def sample_transformed_params(n: int, search_space: SearchSpace) -> np.ndarray:
+    dim = search_space.param_type.shape[0]
+    param_types = search_space.param_type
+    bounds = search_space.bounds
+    steps = search_space.step
+    qmc_engine = scipy.stats.qmc.Sobol(dim, scramble=True)
+    param_values = qmc_engine.random(n)
+    for i in range(dim):
+        if param_types[i] == ParamType.CATEGORICAL:
+            param_values[:, i] = np.floor(param_values[:, i] * bounds[i, 1])
+        elif steps[i] != 0.0:
+            param_values[:, i] = round_one_transformed_param(
+                param_values[:, i], param_types[i], (bounds[i, 0], bounds[i, 1]), steps[i]
+            )
+    return param_values
+
+
+def get_search_space_and_transformed_params(
+    trials: list[optuna.trial.FrozenTrial],
+    optuna_search_space: dict[str, optuna.distributions.BaseDistribution],
+) -> tuple[SearchSpace, np.ndarray]:
+    param_type = np.zeros(len(optuna_search_space), dtype=np.int64)
```

Comment:
I think list is better because we do not take an integer inside the array anymore.

Reply:
Sorry, this is also an artifact of trying to use `numba`.
`numba` is very slow to process Python lists, but it is fast to process `numpy.ndarray`. We want to use `numpy.ndarray` as much as possible.

Reply:
The `param_type` field used in `SearchSpace` class was passed to a `numba` compiled function, I'd like to stick to the `np.ndarray` implementation if this does not change readability very much.